### ISSUE TITLE
BigQuery column ordering change causes destroy of table

### ISSUE THREAD

**Issue by daisy1754** [hashicorp/terraform-provider-google#8337]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

Terraform v0.12.29

### Affected Resource(s)

google_bigquery_table

### Terraform Configuration Files

```tf
resource "google_bigquery_table" "customers_table" {
  dataset_id = google_bigquery_dataset.business.dataset_id
  table_id   = "customers"

  time_partitioning {
    type                     = "DAY"
    field                    = "timestamp"
    require_partition_filter = false
  }

  schema = file("../schemas/customers.schema")

}
```

schema
```
[
 {
  "name": "customer",
  "type": "RECORD",
  "mode": "NULLABLE",
  "fields": [
   {
    "name": "id",
    "type": "STRING",
    "mode": "NULLABLE"
   },
   {
    "name": "created_at",
    "type": "STRING",
    "mode": "NULLABLE",
   },
   {
    "name": "some_complex_object",
    "type": "RECORD",
    "mode": "REPEATED",
    "fields": [
      {
        "name": "some_field",
        "type": "STRING",
        "mode": "NULLABLE",
      }
    ]
   }
  ]
 },
 {
  "name": "request_id",
  "type": "STRING",
  "mode": "NULLABLE"
 }
]
```

Issue happens when we reorder (we autogenerate schame and something changed in auto generated schema) like below:
```
[
 {
  "name": "customer",
  "type": "RECORD",
  "mode": "NULLABLE",
  "fields": [
   {
    "name": "id",
    "type": "STRING",
    "mode": "NULLABLE"
   },
   {
    "name": "some_complex_object",
    "type": "RECORD",
    "mode": "REPEATED",
    "fields": [
      {
        "name": "some_field",
        "type": "STRING",
        "mode": "NULLABLE",
      }
    ]
   }
  ]
 },
 {
  "name": "request_id",
  "type": "STRING",
  "mode": "NULLABLE"
 },
   {
    "name": "created_at",
    "type": "STRING",
    "mode": "NULLABLE",
   }
]
```

### Expected Behavior

I should see no diff when bigquery column order changes in schema file

### Actual Behavior

I saw diff in older version of terraform-provider-google (and I was just ignoring). 
After I update plugin from 3.25.0 to 3.54.0, now terraform try to destroy the table, ie
```
Error: Instance cannot be destroyed

  on ../modules/my.tf line 13:
  13: resource "google_bigquery_table" "customers_table" {

Resource module.my.google_bigquery_table.customers_table has
lifecycle.prevent_destroy set, but the plan calls for this resource to be
destroyed. To avoid this error and continue with the plan, either disable
lifecycle.prevent_destroy or reduce the scope of the plan using the -target
flag.
```

### Steps to Reproduce

1. `terraform apply` with old schema 
2. change order of column in schema
3. `terraform apply` with new schema

### References

* #0000

**Comment by venkykuberan** [comment]
@daisy1754 can you please add the plan output here

**Comment by rileykarson** [comment]
See https://github.com/hashicorp/terraform-provider-google/pull/8325/files where we've reverted the behaviour causing Terraform to attempt to destroy the table. This release should go out midday (PST) Monday.

**Comment by daisy1754** [comment]
@rileykarson @venkykuberan can you reopen this? even after revert this would cause permadiff (at least it did with 3.25.0)

**Comment by daisy1754** [comment]
@venkykuberan I don't have diff handy but it was essentially trying to delete created_at field from certain position and trying to add created_at to certain position

**Comment by rileykarson** [comment]
Sorry, there's a whole network of bugs here and I probably should have done some context links. From my perspective the new part about this bug is the "ordering changes causes destroy" part based on a few other pieces of context below.

It was true for a long time that the provider didn't take reorderings into consideration: https://github.com/hashicorp/terraform-provider-google/issues/4143. I believe it's expected that `3.25.0` doesn't do so.

That behaviour was resolved in https://github.com/GoogleCloudPlatform/magic-modules/pull/3751 which was released in `3.32.0`. At this point the provider should have taken ordering properly into account. Note that there was an edge case issue introduced (https://github.com/hashicorp/terraform-provider-google/issues/7193) fixing in `3.39.0`.

However while fixing another minor bug (https://github.com/hashicorp/terraform-provider-google/issues/7979) a regression was introduced where ordering changes were no longer properly taken into account. That's captured in captured in https://github.com/hashicorp/terraform-provider-google/issues/8282.

That regression was innocuous until the recreation behaviour introduced in `3.54.0`. The `3.53.0` regression caused Terraform to believe that there were modifications to the table's schema and the `3.54.0` change caused it to believe they required it to recreate the cluster.

Where we are now is:

- We've reverted the recreation behaviour for `3.55.0` however the `3.53.0` regression will still be present.
- We've additionally [created a fix for `3.56.0`](https://github.com/GoogleCloudPlatform/magic-modules/pull/4440) where we _believe_ that we're able to both introduce the `3.54.0` behaviour and fix the `3.53.0` regression that caused diffs when schema fields were reordered.

---

tl;dr: The permadiff in earlier versions was captured in https://github.com/hashicorp/terraform-provider-google/issues/4143 and the permadiff in `3.53.0+` is captured in https://github.com/hashicorp/terraform-provider-google/issues/8282. Both are believed to be resolved on our development branch, and we expect this to be working in `3.56.0` on Feb 8th.

**Comment by daisy1754** [comment]
thanks for clarification!

**Comment by stephanepham-sib** [comment]
Hello,
We still can't update the schema in version `v3.58.0`, is it expected?
I'm facing the same issue when Terraform tries to recreate my table.

Regards,
